core/src/main/java/org/apache/iceberg/stats/StatsReader.java

```diff
@@ -0,0 +1,188 @@
+    return knownFileMetadata;
+  }
+
+  private Set<Flag> decodeFlags(byte[] footer, int footerStructOffset) {
+    EnumSet<Flag> flags = EnumSet.noneOf(Flag.class);
+    for (int byteNumber = 0; byteNumber < StatsFormat.FOOTER_STRUCT_FLAGS_LENGTH; byteNumber++) {
+      int flagByte =
+          Byte.toUnsignedInt(footer[footerStructOffset + StatsFormat.FOOTER_STRUCT_FLAGS_OFFSET + byteNumber]);
+      int bitNumber = 0;
+      while (flagByte != 0) {
+        if ((flagByte & 0x1) != 0) {
+          Flag flag = Flag.fromBit(byteNumber, bitNumber);
+          Preconditions.checkState(flag != null, "Unknown flag byte %s and bit %s set", byteNumber, bitNumber);
+          flags.add(flag);
+        }
+        flagByte = flagByte >> 1;
+        bitNumber++;
+      }
+    }
+    return flags;
+  }
+
+  public Iterable<Pair<BlobMetadata, ByteBuffer>> readAll(List<BlobMetadata> blobs) {
+    if (blobs.isEmpty()) {
+      return ImmutableList.of();
+    }
+
+    // TODO inspect blob offsets and coalesce read regions close to each other
+
+    return () -> blobs.stream()
+        .sorted(Comparator.comparingLong(BlobMetadata::offset))
+        .map((BlobMetadata blobMetadata) -> {
+          try {
+            input.seek(blobMetadata.offset());
+            byte[] bytes = new byte[Math.toIntExact(blobMetadata.length())];
+            ByteStreams.readFully(input, bytes);
+            ByteBuffer rawData = ByteBuffer.wrap(bytes);
+            StatsCompressionCodec codec = StatsCompressionCodec.forName(blobMetadata.compressionCodec());
+            ByteBuffer data = StatsFormat.decompress(codec, rawData);
+            return Pair.of(blobMetadata, data);
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        })
+        .iterator();
+  }
+
+  private static void checkMagic(byte[] data, int offset) {
+    byte[] read = Arrays.copyOfRange(data, offset, offset + MAGIC.length);
+    if (!Arrays.equals(read, MAGIC)) {
+      throw new IllegalStateException(String.format(
+          "Invalid file: expected magic at offset %s: %s, but got %s",
+          offset, Arrays.toString(MAGIC), Arrays.toString(read)));
+    }
+  }
+
+  private int footerSize() throws IOException {
+    if (knownFooterSize == null) {
+      if (fileSize < StatsFormat.FOOTER_STRUCT_LENGTH) {
+        throw new IllegalStateException(String.format(
+            "Invalid file: file length %s is less tha minimal length of the footer tail %s",
+            fileSize,
+            StatsFormat.FOOTER_STRUCT_LENGTH));
+      }
+      byte[] footerStruct = readInput(fileSize - StatsFormat.FOOTER_STRUCT_LENGTH, StatsFormat.FOOTER_STRUCT_LENGTH);
+      checkMagic(footerStruct, StatsFormat.FOOTER_STRUCT_MAGIC_OFFSET);
+
+      int footerPayloadSize = StatsFormat.readIntegerLittleEndian(
+          footerStruct, StatsFormat.FOOTER_STRUCT_PAYLOAD_SIZE_OFFSET);
+      knownFooterSize = 4 /* magic */ + footerPayloadSize + StatsFormat.FOOTER_STRUCT_LENGTH;
```

Comment:
Rather than "magic" I think it is more clear to say "Footer start magic". I'd also make this a constant that comes from the magic bytes so we don't have to find hard-coded instances.

Reply:
done